dataset link https://drive.google.com/file/u/2/d/1hEL7PFv7_TRjmdWSp4l9_bh8FTA5fl5p/view?usp=sharing

In [ ]:
# import library
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
data = pd.read_excel('/content/drive/My Drive/DATASET-LATIHAN/online_retail.xlsx')

Mounted at /content/drive


In [ ]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [ ]:
data.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [ ]:
data.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [ ]:
# cleaning data

#1. menghapus ruang ekstra dalam deskripsi
data['Description'] = data['Description'].str.strip()

#2. menghapus baris tanpa nomor faktur (InvoiceNo)
data.dropna(axis=0, subset= ['InvoiceNo'], inplace=True)
data['InvoiceNo'] = data['InvoiceNo'].astype(str)

#3. menghapus semua transaksi yang dilakukan secara kredit
data = data[~data['InvoiceNo'].str.contains('C')]

In [ ]:
# memisahkan data sesuai dengan wilayah transaksi
# prancis
basket_france = (data[data['Country'] == 'France'].groupby(['InvoiceNo','Description'])['Quantity'].sum().unstack().reset_index().fillna(0).set_index('InvoiceNo'))

# portugal
basket_portugal = (data[data['Country'] == 'Portugal'].groupby(['InvoiceNo','Description'])['Quantity'].sum().unstack().reset_index().fillna(0).set_index('InvoiceNo'))

# swedia
basket_swedia = (data[data['Country'] == 'Sweden'].groupby(['InvoiceNo','Description'])['Quantity'].sum().unstack().reset_index().fillna(0).set_index('InvoiceNo'))
pd.set_option('display.max_columns', None)

In [ ]:
basket_swedia

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
538848                                 0.0   
539338                                 0.0   
540040                                 0.0   
542428                                 0.0   
542911                                 0.0   
546161                                 0.0   
546530                                 0.0   
547645                                 0.0   
550277                                 0.0   
551278                                 0.0   
552717                                 0.0   
552957                                 0.0   
553674                                 0.0   
555403                                 0.0   
556194                                 0.0   
557939                                 0.0   
559145                                 0.0   
561890                                 0.0   
561902                                 0.0   
562285                                 0.0   
562600                                 0.0   
562650                                 0.0   
563443                                 0.0   
564539                                 0.0   
565287                                 0.0   
566494                                 0.0   
569018                                 0.0   
571191                                 0.0   
571318                                 0.0   
573333                                 0.0   
574690                                24.0   
575736                                 0.0   
576083                                 0.0   
576519                                 0.0   
578089                                 0.0   
580704                                 0.0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
538848                                       0.0                          0.0   
539338                                       0.0                          0.0   
540040                                       0.0                          0.0   
542428                                       0.0                          0.0   
542911                                       0.0                          0.0   
546161                                       0.0                          0.0   
546530                                     768.0                        384.0   
547645                                       0.0                          0.0   
550277                                       0.0                          0.0   
551278                                       0.0                          0.0   
552717                                       0.0                          0.0   
552957                                       0.0                          0.0   
553674                                       0.0                          0.0   
555403                                       0.0                          0.0   
556194                                       0.0                          0.0   
557939                                       0.0                          0.0   
559145                                       0.0                          0.0   
561890                                       0.0                          0.0   
561902                                       0.0                          0.0   
562285                                       0.0                          0.0   
562600                                       0.0                         24.0   
562650                                       0.0                          0.0   
563443                                       0.0                          0.0   
564539                                       0.0                          0.0   
565287                                       0.0                          0.0   
566494                                       0.0                

jika dilihat diatas pada invoice 546530
muncul angka 768.0 itu berarti barang dibeli sebanyak 768 kali

- tapi didalam asosiasi kita tidak perlu jumlahnya berapapun, tetap akan dihitungnya 1
- karena akan dibuat boolean muncul / tidak muncul

In [ ]:
basket_swedia.shape

(36, 261)

jadi totalnya pembeli di swedia ada 36 orang dan jumlah item/barang ada 261

In [ ]:
# encoding
def hot_encode(x):
  if(x <=0):
    return 0
  if(x >=1):
    return 1

In [ ]:
# encode kan ke tiap tiap wilayah

basket_encoded = basket_france.applymap(hot_encode)
basket_france = basket_encoded

basket_encoded = basket_portugal.applymap(hot_encode)
basket_portugal = basket_encoded

basket_encoded = basket_swedia.applymap(hot_encode)
basket_swedia = basket_encoded

In [ ]:
basket_swedia

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
538848                                   0   
539338                                   0   
540040                                   0   
542428                                   0   
542911                                   0   
546161                                   0   
546530                                   0   
547645                                   0   
550277                                   0   
551278                                   0   
552717                                   0   
552957                                   0   
553674                                   0   
555403                                   0   
556194                                   0   
557939                                   0   
559145                                   0   
561890                                   0   
561902                                   0   
562285                                   0   
562600                                   0   
562650                                   0   
563443                                   0   
564539                                   0   
565287                                   0   
566494                                   0   
569018                                   0   
571191                                   0   
571318                                   0   
573333                                   0   
574690                                   1   
575736                                   0   
576083                                   0   
576519                                   0   
578089                                   0   
580704                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
538848                                         0                            0   
539338                                         0                            0   
540040                                         0                            0   
542428                                         0                            0   
542911                                         0                            0   
546161                                         0                            0   
546530                                         1                            1   
547645                                         0                            0   
550277                                         0                            0   
551278                                         0                            0   
552717                                         0                            0   
552957                                         0                            0   
553674                                         0                            0   
555403                                         0                            0   
556194                                         0                            0   
557939                                         0                            0   
559145                                         0                            0   
561890                                         0                            0   
561902                                         0                            0   
562285                                         0                            0   
562600                                         0                            1   
562650                                         0                            0   
563443                                         0                            0   
564539                                         0                            0   
565287                                         0                            0   
566494                                         0                

# membangun model

In [ ]:
# membangun model dan menganalisis hasilnya

# 1. prancis

# minimal suportnya 0.05
frq_items = apriori(basket_france, min_support= 0.05, use_colnames = True)

# mengumpulkan aturan yang disimpu;kan dalam dataframe
rules = association_rules(frq_items, metric='lift', min_threshold =1)
rules = rules.sort_values(['confidence','lift'], ascending= [False, False])
rules.head()

antecedents  \
45                        (JUMBO BAG WOODLAND ANIMALS)   
260  (PLASTERS IN TIN CIRCUS PARADE, RED TOADSTOOL ...   
271  (RED TOADSTOOL LED NIGHT LIGHT, PLASTERS IN TI...   
302  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
300  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...   

                         consequents  antecedent support  consequent support  \
45                         (POSTAGE)            0.076531            0.765306   
260                        (POSTAGE)            0.051020            0.765306   
271                        (POSTAGE)            0.053571            0.765306   
302  (SET/6 RED SPOTTY PAPER PLATES)            0.102041            0.127551   
300    (SET/6 RED SPOTTY PAPER CUPS)            0.102041            0.137755   

      support  confidence      lift  leverage  conviction  zhangs_metric  
45   0.076531       1.000  1.306667  0.017961         inf       0.254144  
260  0.051020       1.000  1.306667  0.011974         inf       0.247312  
271  0.053571       1.000  1.306667  0.012573         inf       0.247978  
302  0.099490       0.975  7.644000  0.086474   34.897959       0.967949  
300  0.099490       0.975  7.077778  0.085433   34.489796       0.956294

antecedence itu anggap saja A
consequents itu anggap saja B
jadi jika orang membeli jumbo bag woodland animals maka ia membeli postage

In [ ]:
# portugal

frq_items = apriori(basket_portugal, min_support= 0.05, use_colnames = True)

# mengumpulkan aturan yang disimpu;kan dalam dataframe
rules = association_rules(frq_items, metric='lift', min_threshold =1)
rules = rules.sort_values(['confidence','lift'], ascending= [False, False])
rules.head()

antecedents                          consequents  \
1170  (SET 12 COLOUR PENCILS DOLLY GIRL)     (SET 12 COLOUR PENCILS SPACEBOY)   
1171    (SET 12 COLOUR PENCILS SPACEBOY)   (SET 12 COLOUR PENCILS DOLLY GIRL)   
1172  (SET 12 COLOUR PENCILS DOLLY GIRL)   (SET OF 4 KNICK KNACK TINS LONDON)   
1173  (SET OF 4 KNICK KNACK TINS LONDON)   (SET 12 COLOUR PENCILS DOLLY GIRL)   
1174  (SET 12 COLOUR PENCILS DOLLY GIRL)  (SET OF 4 KNICK KNACK TINS POPPIES)   

      antecedent support  consequent support   support  confidence       lift  \
1170            0.051724            0.051724  0.051724         1.0  19.333333   
1171            0.051724            0.051724  0.051724         1.0  19.333333   
1172            0.051724            0.051724  0.051724         1.0  19.333333   
1173            0.051724            0.051724  0.051724         1.0  19.333333   
1174            0.051724            0.051724  0.051724         1.0  19.333333   

      leverage  conviction  zhangs_metric  
1170  0.049049         inf            1.0  
1171  0.049049         inf            1.0  
1172  0.049049         inf            1.0  
1173  0.049049         inf            1.0  
1174  0.049049         inf            1.0

## insight
- antecedents: Ini adalah kumpulan item yang ada sebelum hasil yang diinginkan dalam aturan asosiasi. Dalam contoh yang Anda berikan, antecedents adalah '(SET 12 COLOUR PENCILS DOLLY GIRL)'.

- consequents: Ini adalah item atau kumpulan item yang dihasilkan setelah item atau kumpulan item dalam antecedents dalam aturan asosiasi. Dalam contoh yang Anda berikan, consequents adalah '(SET 12 COLOUR PENCILS SPACEBOY)'.

- antecedent support: Ini adalah proporsi transaksi yang mengandung antecedents dalam keseluruhan dataset. Dalam contoh yang Anda berikan, antecedent support adalah 0.051724, yang berarti sekitar 5.17% dari semua transaksi mengandung '(SET 12 COLOUR PENCILS DOLLY GIRL)'.

- consequent support: Ini adalah proporsi transaksi yang mengandung consequents dalam keseluruhan dataset. Dalam contoh yang Anda berikan, consequent support adalah 0.051724, yang berarti sekitar 5.17% dari semua transaksi mengandung '(SET 12 COLOUR PENCILS SPACEBOY)'.

- support: Ini adalah proporsi transaksi yang mengandung baik antecedents maupun consequents dalam keseluruhan dataset. Dalam contoh yang Anda berikan, support adalah 0.051724, yang berarti sekitar 5.17% dari semua transaksi mengandung kedua item tersebut.

- confidence: Ini adalah seberapa sering aturan ini ditemukan benar. Dalam contoh yang Anda berikan, confidence adalah 1.0, yang berarti setiap kali '(SET 12 COLOUR PENCILS DOLLY GIRL)' dibeli, '(SET 12 COLOUR PENCILS SPACEBOY)' juga dibeli.

In [ ]:
# swedia

frq_items = apriori(basket_swedia, min_support= 0.05, use_colnames = True)

# mengumpulkan aturan yang disimpu;kan dalam dataframe
rules = association_rules(frq_items, metric='lift', min_threshold =1)
rules = rules.sort_values(['confidence','lift'], ascending= [False, False])
rules.tail()

antecedents                                        consequents  \
25485   (POSTAGE)  (CHILDRENS CUTLERY DOLLY GIRL, CHILDRENS CUTLE...   
202     (POSTAGE)                         (CUPCAKE LACE PAPER SET 6)   
385     (POSTAGE)                    (MINI PLAYING CARDS DOLLY GIRL)   
482     (POSTAGE)               (ROUND SNACK BOXES SET OF4 WOODLAND)   
3571    (POSTAGE)  (ROUND SNACK BOXES SET OF4 WOODLAND, ROUND SNA...   

       antecedent support  consequent support   support  confidence      lift  \
25485            0.611111            0.055556  0.055556    0.090909  1.636364   
202              0.611111            0.083333  0.055556    0.090909  1.090909   
385              0.611111            0.083333  0.055556    0.090909  1.090909   
482              0.611111            0.083333  0.055556    0.090909  1.090909   
3571             0.611111            0.083333  0.055556    0.090909  1.090909   

       leverage  conviction  zhangs_metric  
25485  0.021605    1.038889       1.000000  
202    0.004630    1.008333       0.214286  
385    0.004630    1.008333       0.214286  
482    0.004630    1.008333       0.214286  
3571   0.004630    1.008333       0.214286

In [ ]:
rules.shape

(104608, 10)